compiler/sem/lambdalifting.nim

```diff
@@ -558,270 +533,330 @@ proc getUpViaParam(g: ModuleGraph; owner: PSym): PNode =
+    result.body = liftCapturedVars(body, g, idgen, initLiftingPass(d, param))
+    result.env = param
+  elif body.kind != nkEmpty:
+    assert fn.typ.callConv != ccClosure or getEnvParam(fn) != nil,
+      "missing environment parameter"
+
+    var d = initDetectionPass(g, fn, idgen)
+    detectCapturedVars(body, fn, d)
+
+    if d.requireUp or d.accessOuter or d.closureProcCalled or
+       d.capturedVars.len > 0:
+      let t = produceEnvType(d, idgen, fn, fn.info)
+      prepareInnerRoutines(d, idgen, t, fn.info)
+
+      let
+        c = initLiftingPass(d, newEnvVar(g.cache, fn, t, fn.info, idgen))
+        transformed = liftCapturedVars(body, g, idgen, c)
+
+      # XXX: if only the outer environment needs to be passed to inner
+      #      routines and nothing in `fn` itself needs to be lifted,
+      #      then we don't need a dedicated environment that stores just
+      #      the up reference
+      if d.requireUp or d.closureProcCalled or d.capturedVars.len > 0:
+        result.body = rawClosureCreation(g, idgen, c, body.info)
+        result.body.add transformed
+      else:
+        result.body = transformed
+
+      result.env = c.envSym
     else:
-      result = body
-    #if fn.name.s == "get2":
-    #  echo "had something to do ", d.somethingToDo
-    #  echo renderTree(result, {renderIds})
+      # nothing to do
+      result = (body, nil)
+  else:
+    # a routine prootype or otherwise empty routine -> nothing to do
+    result = (body, nil)
 
 proc liftLambdasForTopLevel*(module: PSym, body: PNode): PNode =
   # XXX implement it properly
   result = body
 
+proc liftIterSym*(g: ModuleGraph; n: PNode; idgen: IdGenerator; owner, currEnv: PSym): PNode =
+  ## Transforms  ``(iter)``  to  ``(let env = newClosure[iter](); (iter, env))``.
+  ## This cannot happen as part of ``liftLambdas``, as the iterators's
+  ## environment type is not available at that point.
+  let iter = n.sym
+  assert iter.isIterator
+
+  let
+    hp = getHiddenParam(g, iter)
+    envTyp = hp.typ # the environment's ``ref`` type
+    v = newSym(skLet, getIdent(g.cache, envName), nextSymId(idgen), owner, n.info)
+
+  v.typ = envTyp
+  incl(v.flags, sfShadowed)
+
+  let vnode = newSymNode(v)
+
+  result = newNodeIT(nkStmtListExpr, n.info, iter.typ)
+  result.add newTreeI(nkLetSection, n.info,
+    [newIdentDefs(vnode, newObjConstr(envTyp, n.info))])
+
+  let upField = lookupInRecord(envTyp.base.n, getIdent(g.cache, upName))
+  if upField != nil:
+    # the iterator has an 'up' field, and we have to initialize it here
+    let access = accessEnv(currEnv, upField.typ, n.info, g)
+    result.add(newAsgnStmt(rawIndirectAccess(vnode, upField, n.info), access, n.info))
+
+  result.add makeClosure(g, idgen, iter, vnode, n.info)
+
+proc ensureEnvParam*(graph: ModuleGraph, idgen: IdGenerator, prc: PSym) =
+  ## Problem: top-level anonymous expression can explicitly use the .closure
+  ## calling convention (which doesn't make sense, really). All .closure
```

Comment:
Hm, I guess that depends on what we want the `.closure` pragma to mean. Right now, at least for procedures, it acts as a calling convention indicating that the procedure has a hidden parameter through which captured state is accessed.

Since all locations explicitly defined outside of routines are currently globals, they don't need to be captured, so top-level `.closure` procedures don't make sense, as there's nothing that they can possibly capture. In fact, semantic analysis currently reports an error for all top-level named procedures that explicitly use the `.closure` calling.

If we want locations defined outside of top-level to be treated as *locals* (which I think would be a good idea, at least for locations defined in nested scopes), then top-level `.closure` would indeed make sense.